### PR TITLE
feat: add polyfill support for Intl NumberFormat.formatToParts

### PR DESCRIPTION
**Pull request by RgnDunes**
# Description

This PR adds polyfill support for Intl NumberFormat.formatToParts

## Changes Made

List the main changes made in this pull request.

| Title                  | --- |
| ---------------------- | --- |
| JIRA link              | https://razorpay.atlassian.net/browse/ATLAS-84  |
| Slack thread (if any)  | NA  |
| Product spec           | NA  |
| Tech spec/One-pager    | NA  |
| Bundle Size Difference | NA  |
| Fixes Issue            | NA  |

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] This change is a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Is your change backward compatible ?
- [ ] Tested on major browsers (Chrome, Firefox, Safari, IE) ?
- [ ] Tested in a consumer application(s) ?

## Additional Notes

Any additional information that would be helpful for the reviewer.

## Checklist:

- [x] Add Jira ID(s) in PR title and in the description?
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My code is written in TypeScript and follows the coding standards of the project.
- [ ] I have added relevant documentation and updated the README if necessary.
- [x] My commit messages are clear and follow the project's commit message conventions.
- [ ] Is any external library added?
- [x] My changes do not introduce any new runtime errors or warnings.
- [x] All relevant unit tests have been added/updated in my PR for the change.
- [ ] Any Screenshots (mobile & desktop) required for PR? If yes, have you added the respective screenshots ?
- [ ] Any manual dev testing done by you on beta/func environment?
- [ ] Reviewer added (SLA 2 days)

## Reviewer Checklist

- [ ] Sufficient QA/Dev-Testing is done with proof (test cases list)
- [ ] Ensure that the change in bundle size falls within the acceptable range.

---

## PR Title Format

Format: `<type>: <subject>`

Types can be as follows:

- `feat`: (new feature for the user, not a new feature for build script)
- `fix`: (bug fix for the user, not a fix to a build script)
- `docs`: (changes to the documentation)
- `style`: (formatting, missing semi colons, etc; no production code change)
- `refactor`: (refactoring production code, eg. renaming a variable)
- `test`: (adding missing tests, refactoring tests; no production code change)
- `chore`: (updating grunt tasks etc; no production code change)

### CODE REVIEW

File: src/modules/.internal/polyfills/intlFormatToPartsPolyfill.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  return parts.map((part) => ({
+    type: part.type as Intl.NumberFormatPartTypes, // Casting the custom type to the standard type
+    value: part.value,
+  }));
+}
+
+// Export a default function for the Intl polyfill
+export default function intlFormatToPartsPolyfill() {
+  // Check if the formatToParts method is not defined
+  if (!Intl.NumberFormat.prototype.formatToParts) {
```
Comment (tarun-khanna) [inline src/modules/.internal/polyfills/intlFormatToPartsPolyfill.ts:31]: As far as I can see, `intlFormatToPartsPolyfill` will be triggered even if `Intl` is `undefined`. Wouldn't this throw an error in the condition here ?

File: src/modules/.internal/polyfills/index.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const loadPolyfills = async () => {
+  // Check if the Intl object and Intl.NumberFormat function are available
```
Comment (tarun-khanna) [inline src/modules/.internal/polyfills/index.ts:2]: What is the criteria on injecting polyfill ? 
By below logic, polyfill gets loaded even if Intl and Intl.NumberFormat are absent

File: rollup.config.mjs [hunk 3]
```diff
@@ -67,6 +70,7 @@ export default [
       file: 'lib/esm/index.min.js',
       format: 'es',
       sourcemap: true,
+      inlineDynamicImports: true,
```
[the remark this answers is not in the document]
Comment (tarun-khanna) [reply rollup.config.mjs:73]: Is this the reason for increase in bundle size too in dangerjs report ?

**Review by tarun-khanna — CHANGES_REQUESTED** [review]

File: rollup.config.mjs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RgnDunes) [reply rollup.config.mjs:73]: This is added since for polyfills we wanted dynamic imports.

[the remark this answers is not in the document]
Comment (RgnDunes) [reply rollup.config.mjs:73]: > Is this the reason for increase in bundle size too in dangerjs report ?

Yes

File: src/modules/.internal/polyfills/index.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  if (
+    !(
+      typeof Intl !== 'undefined' &&
```
[the remark this answers is not in the document]
Comment (tarun-khanna) [reply src/modules/.internal/polyfills/index.ts:5]:
```
typeof Intl !== 'undefined' && typeof Intl.NumberFormat === 'function' && typeof Intl.NumberFormat().formatToParts !== 'function'
```

File: src/modules/.internal/polyfills/index.ts (hunk 2 above)
Comment (tarun-khanna) [reply src/modules/.internal/polyfills/index.ts:2]: the condition needs to be updated here. 
Check if Intl is present and NumberFormat is a function and formatToParts is absent.

**Pull request closed without merging**